frontend/src/lib/v2/CompareUtils.ts

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright 2022 The Kubeflow Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { CompareTableProps, xParentLabel } from 'src/components/CompareTable';
+import { getArtifactName, getExecutionName } from 'src/mlmd/MlmdUtils';
+import { getMetadataValue } from 'src/mlmd/Utils';
+import { ExecutionArtifact, RunArtifact } from 'src/pages/CompareV2';
+import { Value } from 'src/third_party/mlmd';
+import * as jspb from 'google-protobuf';
+import { chain, flatten } from 'lodash';
+
+// Add the scalar metric names and data items.
+const addScalarDataItems = (
+  customProperties: jspb.Map<string, Value>,
+  scalarMetricNames: string[],
+  dataMap: { [key: string]: string },
+  artifactIndex: number,
+) => {
+  for (const entry of customProperties.getEntryList()) {
+    const scalarMetricName: string = entry[0];
+    if (scalarMetricName === 'display_name') {
+      continue;
+    }
+    scalarMetricNames.push(scalarMetricName);
+
+    // The scalar metric name and artifact index create a unique key per data item.
+    const key: string = `${scalarMetricName}-${artifactIndex}`;
```

Comment:
I initially attempted to create an object as the key rather than a `string`, but since objects are compared by reference for equality, the `string` was my next best solution.

Reply:
Sounds good, it is also good to call out in the comment hyphen is not allowed in python variable name, which is why we are using it as separator between metricName and artifactIndex.

Another thinking: How about we define `dataMap` as value being an array?

```
dataMap: { [key: string]: string[]},
```

Key: The key is `scalarMetricName`. 
Value: Value is a list of metadataVaue. Index is `artifactIndex`.

Benefit: You don't need to concat `${scalarMetricName}-${artifactIndex}` for the key of map, and everytime you construct row of data, it is simpler :

```
const rows: string[][] = yLabels.map(yLabel => {
    return dataMap[key];
  });
```

Reply:
So I initially tried to do a variant of this as described at https://github.com/kubeflow/pipelines/pull/8026#discussion_r920354359, but ran into an issue where since not every artifact has every scalar metric, the length of each array would not match the number of artifacts. I would need to know what all of the scalar metrics are at the start, so I can loop through each one and then either fill in the data item or leave a blank space. This would require looping through all of the artifacts first before populating the `dataMap`, which seems like extra avoidable work.

If you were describing a system where the `dataMap` is defined as follows:
```
interface DataMapValue {
  value: string;
  artifactIndex: number;
}

dataMap: { [key:string] : DataMapValue[] }
```
Then I could convert that to the `rows` element; however, it would not be as simple as the code you described, I would need to construct a new array either populating the elements if that a `DataMapValue` with that `artifactIndex` exists or adding a blank space if none does.

I think these solutions end up being more complex than the current solution (I initially tried to code this the way you described, but ran into these issues); however, I may have missed something - please let me know if this is the case! I'm open to changes / simplifications.

Reply:
Thank you for this explanation and for the call! Both were very helpful, and this change makes sense. I broke up the implementation into two commits:
- https://github.com/kubeflow/pipelines/pull/8026/commits/2b5844818aeb973608dfdee7c2db8f45087246c3 made the change you described (with the clarification that each array is of length `xLabels.length`, as we discussed). This includes finding the number of artifacts earlier in the code in `CompareV2.tsx`, so that we have the number of `xLabels` and thus can create an array of the correct size when the `scalarMetricName` is not present in the `dataMap`.
- https://github.com/kubeflow/pipelines/pull/8026/commits/1969757b1d76297be91d6b0ec1267298ec4ce4a5 made a further simplification by removing the use of `lodash`, and instead including the sorting of the `yLabels` directly onto the `dataMap` object. Though I had to slightly expand the `dataMap` values to store the number of artifacts in each list as a `number`, this greatly simplified the `getCompareTableProps` function logic.

The main runtime savings were with the first commit. If we assume `n` is the number of artifacts, and `m` is the number of scalar metrics, then the following is true:
- On my initial attempt, I had a guaranteed runtime of `O(n*m)` in the last function logic, as I looped through every row element whether or not scalar metric data existed there.
- On this second attempt, we have simplified the runtime to `O(m)` of that particular section in the last function logic, as I already have the rows and I simply need to loop through each scalar metric name and get the corresponding row.

The above analysis is not counting the big-O runtime on the other parts of the code - you ultimately have to create a 2-d array of size `n*m` so this is done elsewhere - but since this second attempt does not worsen runtime by big-O elsewhere and simplifies the runtime here, this update should improve code runtime and clarity.

Reply:
Awesome implementation and I like that you have done complexity analysis to compare two approaches! Yes the 2-d array will have to consume `n*m` steps to initialize value anyway, but we can save the space and time during computation because we are reusing value in `dataMap` to construct CompareTable and removing extra step in lodash.